src/Serializer/ItemNormalizer.php

```diff
@@ -76,10 +77,35 @@ private function updateObjectToPopulate(array $data, array &$context): void
             $context[self::OBJECT_TO_POPULATE] = $this->iriConverter->getResourceFromIri((string) $data['id'], $context + ['fetch_data' => true]);
         } catch (InvalidArgumentException) {
             $operation = $this->resourceMetadataCollectionFactory->create($context['resource_class'])->getOperation();
-            // todo: we could guess uri variables with the operation and the data instead of hardcoding id
-            $iri = $this->iriConverter->getIriFromResource($context['resource_class'], UrlGeneratorInterface::ABS_PATH, $operation, ['uri_variables' => ['id' => $data['id']]]);
+            $uriVariables = $this->guessUriVariables($data, $operation, $context);
+            $iri = $this->iriConverter->getIriFromResource($context['resource_class'], UrlGeneratorInterface::ABS_PATH, $operation, ['uri_variables' => $uriVariables]);
 
             $context[self::OBJECT_TO_POPULATE] = $this->iriConverter->getResourceFromIri($iri, ['fetch_data' => true]);
         }
     }
+
+    private function guessUriVariables(array $data, $operation, array $context): array
+    {
+        if (false === isset($context['uri_variables'])) {
+            // @todo Improve the message and check the exception class is correct
+            throw new InvalidArgumentException('Missing key "uri_variables" from the context.');
```

Comment:
Wouldn't it be ok to return ['id' => $data['id']] here instead of just bailing out, to mimic the previous incomplete behaviour?

Reply:
Yes, it is actually a good suggestion 💪 Thank you!